src/Storages/MergeTree/MergeMutateExecutor.cpp

```diff
@@ -0,0 +1,147 @@
+#include <Storages/MergeTree/MergeMutateExecutor.h>
+
+#include <Storages/MergeTree/BackgroundJobsExecutor.h>
+
+
+namespace DB
+{
+
+
+/// This is a RAII class which only decrements metric.
+/// It is added because after all other fixes a bug non-executing merges was occurred again.
+/// Last hypothesis: task was successfully added to pool, however, was not executed because of internal exception in it.
+class ParanoidMetricDecrementor
+{
+public:
+    explicit ParanoidMetricDecrementor(CurrentMetrics::Metric metric_) : metric(metric_) {}
+    void alarm() { is_alarmed = true; }
+    void decrement()
+    {
+        if (is_alarmed.exchange(false))
+        {
+            CurrentMetrics::values[metric]--;
+        }
+    }
+
+    ~ParanoidMetricDecrementor() { decrement(); }
+
+private:
+
+    CurrentMetrics::Metric metric;
+    std::atomic_bool is_alarmed = false;
+};
+
+
+void MergeTreeBackgroundExecutor::removeTasksCorrespondingToStorage(StorageID id)
+{
+    std::lock_guard remove_lock(remove_mutex);
+
+    /// First stop the scheduler thread
+    {
+        std::unique_lock lock(mutex);
+        shutdown_suspend = true;
+        has_tasks.notify_one();
+    }
+
+    scheduler.join();
+
+    /// Remove tasks
+    {
+        std::lock_guard lock(currently_executing_mutex);
+
+        for (auto & [task, future] : currently_executing)
+        {
+            if (task->getStorageID() == id)
+                future.wait();
```

Comment:
Note that we wait with active lock.
I would rather go through currently_executing under lock and remember all futures in a separate list, so then mutex can be released

Reply:
Sorry, I didn't write any comment. But this is important, because a race could occur. Consider we moved some tasks to separate container and release the lock. But other threads from thread pool finish their work and remove this task from a `currently_executing` map, so we could wait a task that is not currently executing.

Reply:
If task was executed, shouldn't we return immediately from `future.wait()`?
We only (probably) need to track task life time.
Maybe use shared_ptr. Or maybe only collect futures, taking ownership.